### PR TITLE
Add support for curvilinear elements when exporting cylindrical and spherical meshes to VTK

### PR DESCRIPTION
**Pull request by pshriwise**
This changeset provides the option to export cylindrical and spherical meshes (with optional datasets) to VTK with `vtkQuadraticHex` elements. While the equations of the element surface don't match the cylindrical/spherical boundaries exactly, they can make a big difference in visualization for coarse meshes.

Below is a visualization of a cylindrical mesh with some dummy data applied. On the left is the linear element representation. The right shows the mesh exported with the curvilinear element option.

Screenshot from 2023-05-24 11-14-27

This capability won't have any effect on how `Regular/RectilinearMesh`'s are exported.

### CODE REVIEW

**Comment by paulromano** [comment]
@pshriwise Looks like there are some conflicts on this branch that need to be resolved

**Comment by pshriwise** [comment]
> @pshriwise Looks like there are some conflicts on this branch that need to be resolved

Thanks! Just took care of those.

File: openmc/mesh.py [hunk 1]
```diff
@@ -217,22 +275,177 @@ def write_data_to_vtk(self, points, filename, datasets, volume_normalization=Tru
...
+        # write curvilinear elements using an unstructured grid
+        else:
+            vtk_grid = self._create_vtk_unstructured_grid()
+            writer = vtk.vtkUnstructuredGridWriter()
+
+        # maintain a list of the datasets as added
+        # to the VTK arrays to ensure they persist
+        # in memory until the file is written
+        if datasets is not None:
+            datasets_out = self._apply_vtk_datasets(vtk_grid, datasets, volume_normalization)
```
Comment (paulromano) [inline openmc/mesh.py:311]: Return value is not used -- is that intentional?

File: openmc/mesh.py [hunk 2]
```diff
@@ -271,18 +488,11 @@ def write_data_to_vtk(self, points, filename, datasets, volume_normalization=Tru
...
 
-        # write the .vtk file
-        writer = vtk.vtkStructuredGridWriter()
-        writer.SetFileName(str(filename))
-        writer.SetInputData(vtk_grid)
-        writer.Write()
-
-        return vtk_grid
-
+        return datasets_out
```
Comment (paulromano) [inline openmc/mesh.py:495]:
```suggestion
        return datasets_out

```

File: openmc/mesh.py [hunk 3]
```diff
@@ -1559,42 +1692,20 @@ def volumes(self):
...
-            points=pts_cartesian,
-            filename=filename,
-            datasets=datasets,
-            volume_normalization=volume_normalization
-        )
+    @staticmethod
+    def _convert_to_cartesian(arr, origin):
+        r_xy = arr[..., 0] * np.sin(arr[..., 1])
+        x = r_xy * np.cos(arr[...,2])
+        y = r_xy * np.sin(arr[...,2])
```
Comment (paulromano) [inline openmc/mesh.py:1703]:
```suggestion
        x = r_xy * np.cos(arr[..., 2])
        y = r_xy * np.sin(arr[..., 2])
```

File: openmc/mesh.py [hunk 4]
```diff
@@ -177,7 +177,63 @@ def vertices(self):
...
+            # compute the midpoints along the first dimension
+            midpoints = grids[dims[0]][:-1] + 0.5 * np.diff(grids[dims[0]])
+
+            coords = (midpoints, grids[dims[1]], grids[dims[2]])
+
+            i_grid, j_grid, k_grid = [coords[dims.index(i)] for i in range(3)]
+
+            # re-use the generate vertices method to create the full mesh grid
+            # transpose to get (i, j, k) ordering of the gridpoints
+            midpoint_grid = StructuredMesh._generate_vertices(i_grid, j_grid, k_grid).T.reshape(-1, 3)
```
Comment (paulromano) [inline openmc/mesh.py:219]:
```suggestion
            midpoint_grid = StructuredMesh._generate_vertices(
                i_grid, j_grid, k_grid).T.reshape(-1, 3)
```

File: openmc/mesh.py [hunk 5]
```diff
@@ -1310,41 +1466,18 @@ def volumes(self):
...
+        return self._convert_to_cartesian(self.vertices.T.reshape(-1, 3), self.origin)
+
+    @staticmethod
+    def _convert_to_cartesian(arr, origin):
+        x = arr[..., 0] * np.cos(arr[..., 1]) + origin[0]
+        y = arr[..., 0] * np.sin(arr[..., 1]) + origin[1]
+        arr[..., 0] = x
+        arr[..., 1] = y
+        arr[..., 2] += origin[2]
+        return arr
```
Comment (paulromano) [inline openmc/mesh.py:1480]:
```suggestion
        return arr

```

File: openmc/mesh.py [hunk 6]
```diff
@@ -870,6 +1050,12 @@ def z_grid(self):
     def _grids(self):
         return (self.x_grid, self.y_grid, self.z_grid)
 
+    @property
+    def cartesian_vertices(self):
+        """Returns vertices in cartesian coordiantes. Identical to ``vertices`` for RegularMesh and RectilinearMesh
```
Comment (paulromano) [inline openmc/mesh.py:1055]: I don't understand the last part -- how can it be identical to `vertices` when it uses `self.vertices` and then transforms it?

File: openmc/mesh.py [hunk 7]
```diff
@@ -363,6 +573,12 @@ def width(self):
                 dims =  self._dimension
                 return [(u - l) / d for u, l, d in zip(us, ls, dims)]
 
+    @property
+    def cartesian_vertices(self):
+        """Returns vertices in cartesian coordiantes. Identical to ``vertices`` for RegularMesh and RectilinearMesh
```
Comment (paulromano) [inline openmc/mesh.py:578]: Same comment before about being identical

File: tests/unit_tests/mesh_to_vtk/test.py [hunk 8]
```diff
@@ -1,69 +1,99 @@
...
+
+# test meshes
+reg_mesh = openmc.RegularMesh()
+reg_mesh.lower_left = (0, 0, 0)
+reg_mesh.upper_right = (20, 50, 50)
+reg_mesh.dimension = (10, 20, 30)
+
+rect_mesh = openmc.RectilinearMesh()
+rect_mesh.x_grid = np.linspace(0, 10, 5)
+rect_mesh.y_grid = np.logspace(np.log10(5), np.log10(20), 10)
```
Comment (paulromano) [inline tests/unit_tests/mesh_to_vtk/test.py:32]:
```suggestion
rect_mesh.y_grid = np.geomspace(5., 20., 10)
```

**Review by paulromano — CHANGES_REQUESTED** [review]
Nicely done with this! I only have a few minor comments and suggestions:

File: openmc/mesh.py (hunk 1 above)
Comment (pshriwise) [reply openmc/mesh.py:311]: Yep. The comment above explains why. If we don't keep the `datasets_out` object that holds the NumPy dataset arrays in memory then VTK will write junk data to the file b/c the VTK arrays in `_apply_vtk_datasets` adopts the memory address of the NumPy arrays (this is not documented in the VTK Python interface to my knowledge btw).

Maybe there's a better way to handle this...

File: openmc/mesh.py (hunk 6 above)
Comment (pshriwise) [reply openmc/mesh.py:1055]: Good point. Not quite identical. I'll address that.

File: tests/unit_tests/mesh_to_vtk/test.py (hunk 8 above)
Comment (pshriwise) [reply tests/unit_tests/mesh_to_vtk/test.py:32]: Oooo neat!

File: openmc/mesh.py (hunk 6 above)
Comment (pshriwise) [reply openmc/mesh.py:1057]: This should probably return the same shape as the ``vertices`` property for consistency. I know @eepeterson has a good rationale for why the shape of the `MeshBase
.vertices` property is what it is. I'll make sure that what I'm adding conforms to that.

Comment (pshriwise) [reply openmc/mesh.py:1058]: Okay I've moved these transposes and reshapes into the VTK methods so these external properties will be consistent on the `StructuredMesh` classes.

File: openmc/mesh.py (hunk 1 above)
Comment (pshriwise) [reply openmc/mesh.py]: I opted to remove the `_apply_vtk_datasets` function. I think it's muddying the water here regarding why datasets_out is returned and held in the `write_data_to_vtk` method.

**Review by paulromano — APPROVED** [review]

**Pull request merged**